nni/experiment/config/exp_config.py

```diff
@@ -0,0 +1,136 @@
+    trial_code_directory: utils.PathLike = '.'  # training service field
+    trial_concurrency: int
+    trial_gpu_number: Optional[int] = None  # training service field
+    max_experiment_duration: Union[str, int, None] = None
+    max_trial_number: Optional[int] = None
+    max_trial_duration: Union[str, int, None] = None
+    nni_manager_ip: Optional[str] = None  # training service field
+    use_annotation: bool = False
+    debug: bool = False
+    log_level: Optional[str] = None
+    experiment_working_directory: utils.PathLike = '~/nni-experiments'
+    tuner_gpu_indices: Union[List[int], int, str, None] = None
+    tuner: Optional[_AlgorithmConfig] = None
+    assessor: Optional[_AlgorithmConfig] = None
+    advisor: Optional[_AlgorithmConfig] = None
+    training_service: Union[TrainingServiceConfig, List[TrainingServiceConfig]]
+    shared_storage: Optional[SharedStorageConfig] = None
+
+    def __init__(self, training_service_platform=None, **kwargs):
+        super().__init__(**kwargs)
+        if training_service_platform is not None:
+            assert utils.is_missing(self.training_service)
+            if isinstance(training_service_platform, list):
+                self.training_service = [utils.training_service_config_factory(ts) for ts in training_service_platform]
+            else:
+                self.training_service = utils.training_service_config_factory(training_service_platform)
+            for algo_type in ['tuner', 'assessor', 'advisor']:
+                # add placeholder items, so users can write `config.tuner.name = 'random'`
+                if getattr(self, algo_type) is None:
+                    setattr(self, algo_type, _AlgorithmConfig(name='_none_'))
+
+    def _canonicalize(self, _parents):
+        self.max_experiment_duration = utils.parse_time(self.max_experiment_duration)
+        self.max_trial_duration = utils.parse_time(self.max_trial_duration)
+        if self.log_level is None:
+            self.log_level = 'debug' if self.debug else 'info'
+        self.tuner_gpu_indices = utils.canonical_gpu_indices(self.tuner_gpu_indices)
+
+        for algo_type in ['tuner', 'assessor', 'advisor']:
+            algo = getattr(self, algo_type)
+            if algo is not None and algo.name == '_none_':
+                setattr(self, algo_type, None)
+
+        super()._canonicalize([self])
+
+    def _validate_canonical(self):
+        super()._validate_canonical()
+
+        space_cnt = (self.search_space is not None) + (self.search_space_file is not None)
+        if self.use_annotation and space_cnt != 0:
+            raise ValueError('ExperimentConfig: search space must not be set when annotation is enabled')
+        if not self.use_annotation and space_cnt != 1:
+            raise ValueError('ExperimentConfig: search_space and search_space_file must be set one')
+
+        if self.search_space_file is not None:
+            self.search_space = yaml.safe_load(open(self.search_space_file))
```

Comment:
does the file need close()?

Reply:
Well I'm pretty sure it won't cause problem to open a file read-only and let GC to close it, but you are right it's better to use `with`.